pgserver: spawn_blocking in compaction

skyzh opened the pull request:
## Problem

## Summary of changes

Compaction is usually a compute-heavy process and might affect other futures running on the thread of the compaction. Therefore, we add `block_in_place` as a temporary solution to avoid blocking other futures on the same thread as compaction in the runtime. As we are migrating towards a fully-async-style pageserver, we can revert this change when everything is async and when we move compaction to a separate runtime.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

hlinnaka reviewed (approved):
@problame you were worried about blocking I/O starving other activity in https://github.com/neondatabase/neon/pull/4220. This PR addresses the write-path rather than reads. These write and fsync operations are much longer than reads of individual pages, so this feels more likely to cause starvation than the reads. Also a much smaller patch :-).

Maybe we need both, but this PR feels more low-hanging fruit than PR 4220.

hlinnaka commented:
par_fsync would be pretty straightforward to convert into async, as an independent part of the bigger migration to async.

koivunej commented:
block_in_place also has a pitfall with block_on, in addition to making the core travel which causes headscratching. Would hope to see another way.

skyzh commented:
... as I originally only thought of asyncify par_fsync and then realize the compaction process has no yield point and we might need to avoid starving in some ways, and I agree that just using `block_in_place` might cause some problems. So I'll probably refactor it as:

* for compaction_l0, run it in a separate thread (as it is compute heavy and might take long time)
* for par_fsync, simply block_on (as par_fsync is already creating new threads and should be relatively fast)

skyzh commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -3040,7 +3042,10 @@ impl Timeline {
             .collect::<Vec<_>>();
-        par_fsync::par_fsync(&all_paths).context("fsync of newly created layer files")?;
+
+        tokio::task::block_in_place(|| {
```
Yes as I suppose fsync should be fast and don’t worth the effort to put it on another thread? Also we are already using rayon thread pool in par_fsync.

Or we can refactor par_fsync to be async and it should be easy.

skyzh replied:
I just realized this PR does not even compile and I've refactored self -> `Arc<Self>` and cs mutex -> OwnedMutexGuard. Should compile now and ready for review.

skyzh replied:
... as `spawn_blocking` requires `'static`.

skyzh replied:
> Tokio already manages thread pool for blocking operations, so not sure of we need to duplicate this logic

Sounds reasonable and now we have two `par_fsync` functions. One for use within async context (which uses tokio thread pool), and another for use within non-async context (i.e., when we spawn_blocking). We can remove the latter one when everything becomes async.

skyzh commented:
Request a final review from @LizardWizzard and I think we can merge this. Thanks for reviewing!

LizardWizzard commented on pageserver/src/tenant/par_fsync.rs:
```diff
@@ -53,3 +49,29 @@ pub fn par_fsync(paths: &[PathBuf]) -> io::Result<()> {
+
+/// Parallel fsync all files. Can be used in non-async context as it is using rayon thread pool.
+pub fn par_fsync(paths: &[PathBuf]) -> io::Result<()> {
+    const PARALLEL_PATH_THRESHOLD: usize = 1;
```
Hm, I know this is pre-existing code, but I think it makes sense to simplify it given this constant (remove the loop, replace the check with `if paths.len() == 1 {...}`

koivunej replied:
Note I doubt we ever use this with `paths.len() == 1`, but with two, layer file and it's parent directory, hence my comment: https://github.com/neondatabase/neon/pull/4265#pullrequestreview-1436645734

skyzh replied:
I do believe this is a problem and I'm going to split fsync into two. First fsync all files, and then the parent directory.

LizardWizzard commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -187,7 +187,7 @@ pub struct Timeline {
     /// This lock is acquired in [`Timeline::gc`], [`Timeline::compact`],
     /// and [`Tenant::delete_timeline`].
-    pub(super) layer_removal_cs: tokio::sync::Mutex<()>,
+    pub(super) layer_removal_cs: Arc<tokio::sync::Mutex<()>>,
```
Lets describe why this needs to be an Arc. As I see this is needed because of spawn blocking as it requires for spawned closure to be 'static, so you cannot pass references to guards inside. This should go away with asyncification

LizardWizzard commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -3511,7 +3523,7 @@ impl Timeline {
         for l in deltas_to_compact {
             layer_names_to_delete.push(l.filename());
-            self.delete_historic_layer(layer_removal_cs, l, &mut updates)?;
+            self.delete_historic_layer(layer_removal_cs.clone(), l, &mut updates)?;
```
NIT: I think using reference instead of an owned Arc should be enough `&layer_removal_cs`. Not really important, just needless clone

skyzh replied:
Sometimes we need to pass this into inner functions so I keep all these functions accepting `Arc<MutexGuard>`, though this one seems like a leaf call 🤣

LizardWizzard replied:
Yep, since this is a leaf call there is no need to clone the Arc once more

LizardWizzard reviewed (approved):
Looks good to me, thanks!

I dont really like growing number of arcs (and nested arcs) but it looks like necessary evil right now. When compaction will be entirely async it should be possible to remove those arcs

skyzh commented on pageserver/src/tenant/par_fsync.rs:
```diff
@@ -19,14 +21,8 @@ fn parallel_worker(paths: &[PathBuf], next_path_idx: &AtomicUsize) -> io::Result
     Ok(())
 }
 
-pub fn par_fsync(paths: &[PathBuf]) -> io::Result<()> {
```
I think we can reuse the async path from spawn_blocking as long as we move fsync out of spawn_blocking of `create_delta_layer`, but I prefer to do this in future PRs.

koivunej replied:
Not sure about how will that happen without Handle::block_on, but let's see it then :) Please remember the deadlock consideration. Resolving.

skyzh replied:
> can we now remove rayon

not yet because IIRC we cannot use async functions inside `spawn_blocking`? once we move fsync out of spawn_blocking we can remove rayon.

> reusing the threads vs. always not reusing the threads

I would prefer reusing threads; once we switch to tokio we can automatically gain the benefit of running everything in tokio worker thread 🤪

But in this case, compaction / flush to L0 are background operations so the overhead does not matter that much. So I don't think we need to make the change right now.

koivunej replied:
On mobile right now but doing a quick code search (probably in your pr branch) shows that we have no more usages for rayon.

Note the other part uses crossveam_utils thread scope, which is superceded by std::thread::scope.

skyzh replied:
using std::thread::scope now

skyzh commented on pageserver/src/tenant/par_fsync.rs:
```diff
@@ -53,3 +49,29 @@ pub fn par_fsync(paths: &[PathBuf]) -> io::Result<()> {
+/// execution thread. Otherwise, we will spawn_blocking and run it in tokio.
+pub async fn par_fsync_async(paths: &[PathBuf]) -> io::Result<()> {
+    const MAX_CONCURRENT_FSYNC: usize = 64;
+    let mut s = stream::iter(paths.to_vec())
+        .map(|path| async move { tokio::task::spawn_blocking(move || fsync_path(&path)).await })
+        .buffer_unordered(MAX_CONCURRENT_FSYNC);
```
This depends on https://github.com/neondatabase/neon/pull/4294 and I'd prefer updating this PR later when the tokio upgrade PR is merged (or we can merge this first). 🤪

koivunej replied:
Merged the tokio upgrade.

skyzh replied:
The comments are applied and ready for review, thanks!

skyzh commented:
Fixed the fsync problem and updated the criteria for no spawning to rayon thread pool to `== 1`. Waiting for review + tokio upgrade PR to be merged. Thanks for the comments!
